File: python/bowline-agent/src/bowline_agent/encode.py
from __future__ import annotations

from typing import Any

from .tools import Tool


def to_anthropic(tools: list[Tool]) -> list[dict[str, Any]]:
    return [
        {"name": t.name, "description": t.description, "input_schema": t.input_schema}
        for t in tools
    ]


def to_openai(tools: list[Tool]) -> list[dict[str, Any]]:
    return [
        {
            "type": "function",
            "function": {
                "name": t.name,
                "description": t.description,
                "parameters": t.input_schema,
            },
        }
        for t in tools
    ]


def to_json_schema(tools: list[Tool]) -> list[dict[str, Any]]:
    return [
        {
            "name": t.name,
            "procedure": t.procedure,
            "description": t.description,
            "inputSchema": t.input_schema,
            "outputSchema": t.output_schema,
            "readOnly": t.read_only,
            "destructive": t.destructive,
            "scopes": sorted(t.scopes),
        }
        for t in tools
    ]

File: packages/agent/src/encode.ts
import type { Tool } from "./tools.js";

export interface AnthropicTool {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
}

export interface OpenAITool {
  type: "function";
  function: {
    name: string;
    description: string;
    parameters: Record<string, unknown>;
  };
}

export interface SchemaTool {
  name: string;
  procedure: string;
  description: string;
  inputSchema: Record<string, unknown>;
  outputSchema: Record<string, unknown>;
  readOnly: boolean;
  destructive: boolean;
  scopes: string[];
}

export function toAnthropic(list: Tool[]): AnthropicTool[] {
  return list.map((t) => ({
    name: t.name,
    description: t.description,
    input_schema: t.inputSchema,
  }));
}

export function toOpenAI(list: Tool[]): OpenAITool[] {
  return list.map((t) => ({
    type: "function",
    function: { name: t.name, description: t.description, parameters: t.inputSchema },
  }));
}

export function toJSONSchema(list: Tool[]): SchemaTool[] {
  return list.map((t) => ({
    name: t.name,
    procedure: t.procedure,
    description: t.description,
    inputSchema: t.inputSchema,
    outputSchema: t.outputSchema,
    readOnly: t.readOnly,
    destructive: t.destructive,
    scopes: [...t.scopes].sort(),
  }));
}
